spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/observation/TracingObservationRegistryCustomizer.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2012-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.autoconfigure.observation;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.micrometer.observation.Observation.GlobalTagsProvider;
+import io.micrometer.observation.ObservationHandler;
+import io.micrometer.observation.ObservationHandler.FirstMatchingCompositeObservationHandler;
+import io.micrometer.observation.ObservationPredicate;
+import io.micrometer.observation.ObservationRegistry;
+import io.micrometer.tracing.handler.TracingObservationHandler;
+
+/**
+ * Customizes a {@link ObservationRegistry} for Micrometer tracing. Works like
+ * {@link NoTracingObservationRegistryCustomizer}, only that
+ * {@link TracingObservationHandler} are grouped into a
+ * {@link FirstMatchingCompositeObservationHandler}.
+ *
+ * This class is used by {@link ObservationAutoConfiguration} when a
+ * {@link TracingObservationHandler} is found on the classpath.
+ *
+ * @author Moritz Halbritter
+ */
+class TracingObservationRegistryCustomizer extends NoTracingObservationRegistryCustomizer {
```

Comment:
Naming is hard, but it feels a little strange that `Tracing…` is a subclass of `NoTracing`. The naming suggests that one is the opposite of the other which, to me at least, means that the inheritance doesn't really fit.

Reply:
Yes, i abused OOP to get the grouping logic right. I refactored the code, it's now using a `ObservationHandlerGrouping` strategy interface. Let me know what you think!